drivers/power/max42500/max42500.c

```diff
@@ -0,0 +1,473 @@
+	if (ret)
+		goto free_addr;
+
+	/* Initialize I2C communication */
+	ret = no_os_i2c_init(&descriptor->comm_desc, &init_param->comm_param);
+	if (ret)
+		goto free_addr;
+
+	/* Check device silicon ID */
+	ret = max42500_reg_read(descriptor, &descriptor->regs[MAX42500_ID]);
+	if (ret)
+		goto free_i2c;
+
+	if (descriptor->regs[MAX42500_ID].value != MAX42500_SILICON_ID) {
+		ret = -ENODEV;
+		goto free_i2c;
+	}
+
+	/* Initially read all registers */
+	for (int8_t reg_id = MAX42500_CONFIG1; reg_id < MAX42500_REG_NO; reg_id++) {
+		ret = max42500_reg_read(descriptor, &descriptor->regs[reg_id]);
+		if (ret)
+			goto free_i2c;
+	}
+
+	/* Configure PEC */
+	ret = max42500_reg_update(descriptor,
+				  MAX42500_CONFIG1,
+				  MAX42500_CONFIG1_PECE_MASK,
+				  init_param->pece);
+	if (ret)
+		goto free_i2c;
+
+	/* Set PEC enable/disable for subsequent register access */
+	descriptor->pece = init_param->pece;
+
+	/* Enable voltage monitor inputs */
+	ret = max42500_reg_update(descriptor,
+				  MAX42500_VMON,
+				  MAX42500_VMON_ALL_VM_MASK,
+				  init_param->vmon_en);
+	if (ret)
+		goto free_i2c;
+
+	/* Enable voltage monitor power-down */
+	ret = max42500_reg_update(descriptor,
+				  MAX42500_VMON,
+				  MAX42500_VMON_VMPD_MASK,
+				  init_param->vmon_vmpd);
+	if (ret)
+		goto free_i2c;
+
+	/* Enable input OV/UV mapping to reset pin */
+	ret = max42500_reg_update(descriptor,
+				  MAX42500_RSTMAP,
+				  MAX42500_RSTMAP_ALL_IN_MASK,
+				  init_param->reset_map);
+	if (ret)
+		goto free_i2c;
+
+	/* Set watchdog mode */
+	ret = max42500_reg_update(descriptor,
+				  MAX42500_WDCDIV,
+				  MAX42500_WDCDIV_SWW_MASK,
+				  init_param->wd_mode);
+	if (ret)
+		goto free_i2c;
+
+	/* Set watchdog clock div */
+	ret = max42500_reg_update(descriptor,
+				  MAX42500_WDCDIV,
+				  MAX42500_WDCDIV_WDIC_MASK,
+				  init_param->wd_cdiv);
+	if (ret)
+		goto free_i2c;
+
+	/* Set watchdog open window */
+	ret = max42500_reg_update(descriptor,
+				  MAX42500_WDCFG1,
+				  MAX42500_WDCFG1_OPN_MASK,
+				  init_param->wd_open);
+	if (ret)
+		goto free_i2c;
+
+	/* Set watchdog close window */
+	ret = max42500_reg_update(descriptor,
+				  MAX42500_WDCFG1,
+				  MAX42500_WDCFG1_CLO_MASK,
+				  init_param->wd_close);
+	if (ret)
+		goto free_i2c;
+
+	/* Set watchdog first update window */
+	ret = max42500_reg_update(descriptor,
+				  MAX42500_WDCFG2,
+				  MAX42500_WDCFG2_1UD_MASK,
+				  init_param->wd_1ud);
+	if (ret)
+		goto free_i2c;
+
+	/* Set watchdog enable */
+	ret = max42500_reg_update(descriptor,
+				  MAX42500_WDCFG2,
+				  MAX42500_WDCFG2_WDEN_MASK,
+				  init_param->wd_en);
+	if (ret)
+		goto free_i2c;
+
+	/* Initialize timer for watchdog */
+	ret = no_os_timer_init(&descriptor->timer_desc, &init_param->timer_param);
+	if (ret)
+		goto free_i2c;
+
+	/* Initialize IRQ controller */
+	ret = no_os_irq_ctrl_init(&descriptor->irq_desc, &init_param->irq_param);
+	if (ret)
+		goto free_timer;
```

Comment:
This watchdog functionality seems to only be half implemented. I assume you want to reset the watchdog based on a timer interrupt? Anyway, I would suggest to let the user decide whether or not they need this (by checking the `timer_param` field for `NULL` for example).

Also, currently, the user of the driver has to register a callback for the timer and enable it in order for this to be useful. If you choose to keep this functionality, I would suggest to add a few parameters or functions in order to do the full interrupt setup.

Reply:
I will remove the watchdog functionality for now.